keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,820 @@
+// Can be either “Filesystem” or “Block”.
+// If left empty, will be treated as “Unknown”.
+// +optional
+SourceVolumeMode *SourceVolumeMode
+...
+```
+
+2. A new annotation to VolumeSnapshotContent objects. The onus is on the 
+backup vendor (via s/w or manually) to add this annotation to the VolumeSnapshotContent
+if they intend to alter the volume mode during a restore operation.
+
+```go
+snapshot.storage.kubernetes.io/allowVolumeModeChange: true
+```
+
+### Changes to Snapshot Controller
+
+There are two cases to consider:
+1. Dynamic Provisioning
+   1. `VolumeSnapshot` is created by the user, with `VolumeSnapshotClass` optionally
+   specified in the spec.
+   2. `VolumeSnapshotContent` is created by the `snapshot-controller` in response to (i).
+   3. `snapshot-controller` populates the `Spec` of the given `VolumeSnapshotContent`.
+      1. With this change, the controller will fetch the `Spec.PersistentVolumeMode`
+      of the `PV` and add that to newly introduced `Spec.SourceVolumeMode` field of
+      the VolumeSnapshotContent to be created.
+2. Static Provisioning
+   `VolumeSnapshotContent` is created by the admin. With this change, the admin will be 
+expected to fill the `Spec.SourceVolumeMode` field appropriately. If left nil, `Unknown`
+mode will be assumed to preserve existing behavior.
+
+### Changes to external-provisioner
+
+This design leverages the access rights of a user on `VolumeSnapshotContents` to
+determine whether the volume mode can be modified during a restore operation.
+The volume mode can be altered if the requesting user has rights to `Update` and
+`Patch` a`VolumeSnapshotContent` (which is a cluster scoped resource).
+The control flow of a restore operation for a `PVC` from a `VolumeSnapshot` will 
+look like below:
+
+1. A user attempts to restore a `PVC` from a `VolumeSnapshot` by specifying the
+`Spec.DataSource` parameter of the `PVC` YAML.
+2. The `external-provisioner` receives a callback to dynamically create the volume.
+As part of the preprocessing steps, it will:
+   1. Get the `Spec.SourceVolumeMode` of the `VolumeSnapshotContent`.
+      1. If `Spec.SourceVolumeMode` doesn't exist or is nil, then continue with
+      volume provisioning to preserve existing behavior.
+   2. Get the `Spec.VolumeMode` of the `PVC` being restored.
+   If they do not match:
+      1. Get all annotations on the `VolumeSnapshotContent` and verify if 
+      `snapshot.storage.kubernetes.io/allowVolumeModeChange: true` exists.
+      If it does not exist, block volume provisioning by returning an error.
+4. In all other cases, let volume provisioning continue.
+
+NOTE: `external-provisioner` maintains a reference to `PVC` and `VolumeSnapshotContent`
+during volume creation. This proposal leverages those references to make additional
+decisions.
+
+### Test Plan
+
+E2E tests will be added for this design, that attempt to restore a volume with
+and without requisite privileges. 
```

Comment:
Can you write the test plan like the following example?
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/177-volume-snapshot#test-plan

Reply:
Done, PTAL!